cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1606,42 +1656,60 @@ class ObjectOutputStream final : public io::OutputStream {
       return Status::OK();
     }
 
-    ARROW_ASSIGN_OR_RAISE(auto client_lock, holder_->Lock());
+    if (is_multipart_created_) {
+      ARROW_ASSIGN_OR_RAISE(auto client_lock, holder_->Lock());
 
-    S3Model::AbortMultipartUploadRequest req;
-    req.SetBucket(ToAwsString(path_.bucket));
-    req.SetKey(ToAwsString(path_.key));
-    req.SetUploadId(upload_id_);
+      S3Model::AbortMultipartUploadRequest req;
+      req.SetBucket(ToAwsString(path_.bucket));
+      req.SetKey(ToAwsString(path_.key));
+      req.SetUploadId(upload_id_);
 
-    auto outcome = client_lock.Move()->AbortMultipartUpload(req);
-    if (!outcome.IsSuccess()) {
-      return ErrorToStatus(
-          std::forward_as_tuple("When aborting multiple part upload for key '", path_.key,
-                                "' in bucket '", path_.bucket, "': "),
-          "AbortMultipartUpload", outcome.GetError());
+      auto outcome = client_lock.Move()->AbortMultipartUpload(req);
+      if (!outcome.IsSuccess()) {
+        return ErrorToStatus(
+            std::forward_as_tuple("When aborting multiple part upload for key '",
+                                  path_.key, "' in bucket '", path_.bucket, "': "),
+            "AbortMultipartUpload", outcome.GetError());
+      }
     }
+
     current_part_.reset();
     holder_ = nullptr;
     closed_ = true;
+
     return Status::OK();
   }
 
   // OutputStream interface
 
+  bool ShouldBeMultipartUpload() { return pos_ > kMultiPartUploadThresholdSize; }
+
+  bool IsMultipartUpload() { return ShouldBeMultipartUpload() || is_multipart_created_; }
+
   Status EnsureReadyToFlushFromClose() {
-    if (current_part_) {
-      // Upload last part
-      RETURN_NOT_OK(CommitCurrentPart());
-    }
+    if (IsMultipartUpload()) {
```

Comment:
Can this actually happens when `part_size_` is large enough but not being uploaded?

Reply:
Can you elaborate more? I think we're calling `EnsureReadyToFlushFromClose` just to flush the data if we close the stream, hence sometimes we still have data in-flight that has to be uploaded (the "last part" which we commit in this branch, or a dummy part if no data was written at all).

Reply:
Just to make it clear, since if `part_size_` is larger than size-limit, it should be switch to multipart during "doWrite"

Reply:
You mean if `current_part_size_` > `kPartUploadSize` can be true at the time we close the stream?
I think that is not possible (and also wasn't possible with the old impl). The invariant when `Close()` is called is that `current_part_size_` < `kPartUploadSize` holds, because if we write to the stream and `current_part_size_` >= `kPartUploadSize` holds, we always directly upload a part in `DoWrite`.

Reply:
Yes, maybe simplify to `has_multi_part_upload_` or other is ok, but this also looks ok to me